src/Chainweb/Pact/PactService.hs

```diff
@@ -509,112 +509,129 @@ syncToFork logger serviceEnv hints forkInfo = do
-                        then
-                            return $
-                                (Checkpointer.Pact5RunnableBlock $ \chainwebPactDb -> do
-                                    (_, pwo, validatedTxs) <-
-                                        Pact.execExistingBlock logger serviceEnv
-                                            (BlockEnv blockCtx chainwebPactDb)
-                                            (CheckablePayload expectedPayloadHash payload)
-                                    -- add payload immediately after executing the block, because this is when we learn it's valid
-                                    liftIO $ addNewPayload
-                                        (_payloadStoreTable $ _psPdb serviceEnv)
-                                        (_evaluationCtxCurrentHeight evalCtx)
-                                        pwo
-                                    return $ (DList.singleton validatedTxs, (_ranked rankedBHash, expectedPayloadHash))
-                                )
-                        else
-                            return $
-                                (Checkpointer.Pact4RunnableBlock $ \blockDbEnv -> do
-                                    (_, pwo) <-
-                                        Pact4.execBlock logger serviceEnv
-                                            (Pact4.BlockEnv blockCtx blockDbEnv)
-                                            (CheckablePayload expectedPayloadHash payload)
-                                    -- add payload immediately after executing the block, because this is when we learn it's valid
-                                    liftIO $ addNewPayload
-                                        (_payloadStoreTable $ _psPdb serviceEnv)
-                                        (_evaluationCtxCurrentHeight evalCtx)
-                                        pwo
-                                    -- don't remove pact 4 txs from the mempool, who cares when we can't make Pact 4 blocks anymore?
-                                    return $ (mempty, (_ranked rankedBHash, expectedPayloadHash))
-                                    )
+          else do
+            -- check if some past block had the target as its parent; if so, that
+            -- means we can rewind to it
+            --
+            -- FIXME: why the parent? Why not the block itself?
+            --
+            latestBlockRewindable <-
+                isJust <$> Checkpointer.lookupBlockHash sql (_latestBlockHash forkInfo._forkInfoTargetState)
+
+            if latestBlockRewindable
+              then do
+                -- we just have to rewind and set the final + safe blocks
+                -- TODO PP: disallow rewinding final?
+                logFunctionText logger Debug $ "pure rewind to " <> brief forkInfo._forkInfoTargetState
+                rewoundTxs <- getRewoundTxs (Parent forkInfo._forkInfoTargetState._consensusStateLatest._syncStateHeight)
```

Comment:
I actually really dislike `OverloadedRecordDot`, can we use lenses here?

Reply:
It's used a lot in that module. It also seems to break HLS in my editor. Yeah, maybe we should replace it by "normal" function composition via ` . `.